File: InsertDeleteGetRandomO(1)-Duplicatesallowed2.py
import random


class RandomizedCollection:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.locs = {}
        self.nums = []

    def insert(self, val):
        """
        Inserts a value to the collection. Returns true if the collection did not already contain the specified element.
        :type val: int
        :rtype: bool
        """
        ret = False
        if val not in self.locs:
            ret = True
            self.locs[val] = []

        self.locs[val].append(len(self.nums))
        self.nums.append((val, len(self.locs[val]) - 1))
        return ret

    def remove(self, val):
        """
        Removes a value from the collection. Returns true if the collection contained the specified element.
        :type val: int
        :rtype: bool
        """
        if val not in self.locs:
            return False
        index = self.locs[val][-1]
        self.locs[val].pop()
        if not self.locs[val]:
            self.locs.pop(val)
        if index < len(self.nums) - 1:
            a, b = self.nums[-1]
            self.nums[index] = self.nums[-1]
            self.locs[a][b] = index
        self.nums.pop()
        return True

    def getRandom(self):
        """
        Get a random element from the collection.
        :rtype: int
        """
        return self.nums[random.randint(0, len(self.nums) - 1)][0]


if __name__ == '__main__':
    obj = RandomizedCollection()
    obj.insert(1)
    obj.remove(1)
    obj.insert(2)
    print(obj.getRandom())

File: InsertDeleteGetRandomO(1)-Duplicatesallowed2.cpp
#include<iostream>
#include<vector>
#include<unordered_map>

using namespace std;

class RandomizedCollection {
private:
	unordered_map<int, vector<int> > locs;
	vector<pair<int, int> > nums;
public:
    /** Initialize your data structure here. */
    RandomizedCollection() {
        
    }
    
    /** Inserts a value to the collection. Returns true if the collection did not already contain the specified element. */
    bool insert(int val) {
    	int ret = false;
        if(locs.find(val) == locs.end()) {
        	locs[val] = vector<int>();
        	ret = true;
		}
		//nums.push_back(val);
		locs[val].push_back(nums.size());
		nums.push_back(make_pair(val, locs[val].size() - 1));
		return ret;
    }
    
    /** Removes a value from the collection. Returns true if the collection contained the specified element. */
    bool remove(int val) {
        if(locs.find(val) == locs.end()) return false;
        int index = locs[val].back();
        locs[val].pop_back();
        if(locs[val].empty()) locs.erase(val);
        
        if(index < nums.size() - 1) {
        	pair<int, int> last = nums.back();
	        //swap(nums[index], nums[nums.size() - 1]);
	        nums[index] = nums.back();
	        locs[last.first][last.second] = index;
		}
        
        nums.pop_back();
        
        return true;
    }
    
    /** Get a random element from the collection. */
    int getRandom() {
        return nums[rand() % nums.size()].first;
    }
};

int main() {
	
	RandomizedCollection *obj = new RandomizedCollection();
	obj->insert(1);
	obj->remove(1);
	obj->insert(1);
	
	cout<<obj->getRandom();
	
	return 0;
	
}
